crates/runtime/src/dataconnector/s3.rs

```diff
@@ -93,74 +97,69 @@ impl DataConnectorFactory for S3 {
     }
 }
 
-#[async_trait]
-impl DataConnector for S3 {
-    fn as_any(&self) -> &dyn Any {
-        self
-    }
-
-    fn get_object_store(
-        &self,
-        dataset: &Dataset,
-    ) -> Option<AnyErrorResult<(Url, Arc<dyn ObjectStore + 'static>)>> {
-        let result: AnyErrorResult<(Url, Arc<dyn ObjectStore + 'static>)> = (|| {
-            let url = dataset.from.clone();
-            let parts = url.clone().replace("s3://", "");
-
-            let bucket = parts
-                .split('/')
-                .next()
-                .ok_or_else(|| MissingForwardSlashSnafu { url: url.clone() }.build())?;
-
-            let mut s3_builder = AmazonS3Builder::new()
-                .with_bucket_name(bucket)
-                .with_allow_http(true);
-
-            if let Some(region) = self.params.get("region") {
-                s3_builder = s3_builder.with_region(region);
-            }
-            if let Some(endpoint) = self.params.get("endpoint") {
-                s3_builder = s3_builder.with_endpoint(endpoint);
-            }
-            if let Some(secret) = &self.secret {
-                if let Some(key) = secret.get("key") {
-                    s3_builder = s3_builder.with_access_key_id(key);
-                };
-                if let Some(secret) = secret.get("secret") {
-                    s3_builder = s3_builder.with_secret_access_key(secret);
-                };
-            } else {
-                s3_builder = s3_builder.with_skip_signature(true);
+impl S3 {
+    fn get_object_store_url(&self, dataset: &Dataset) -> AnyErrorResult<Url> {
+        let mut fragments = vec![];
+        let mut query = form_urlencoded::Serializer::new(String::new());
```

Comment:
```suggestion
        let mut fragment_builder = form_urlencoded::Serializer::new(String::new());
```

This is really a fragment builder